Add a check to detect duplicate k cell declarations

SchmErik opened the pull request:
This pull request adds a check to detect when a `<k>` cell is being declared multiple times.

fixes: https://github.com/runtimeverification/k/issues/2171

Commit:
Check for duplicated K cell declarations

Commit:
Add kompile tests for duplicate cells check

radumereuta commented on kernel/src/main/java/org/kframework/kompile/DefinitionParsing.java:
```diff
@@ -285,6 +285,17 @@ protected Definition resolveConfigBubbles(Definition definition, Module defaultC
+            KCellDeclarations.stream().forEach(k ->
+                errors.add(KEMException.compilerError("Multiple K cell declarations detected. Only one declaration is allowed.", k)));
+        }
+
```
```
configuration <k> "<k>" </k>
```
If you do this at the text level, you will fail for such an example.

But you got the right idea. Below this function you have 
```java
Definition defWithParsedConfigs = ...

// replace config bubbles with the generated syntax and rules
```
You have to write a [K visitor](https://github.com/runtimeverification/k/blob/master/kernel/src/main/java/org/kframework/compile/CloseCells.java#L102) over `defWithParsedConfigs` that looks for `#token("k","#CellName")`

Here is an example configuration
```
configuration <k> $PGM:KItem </k>
              <state> .Map </state>
Configuration(#cells(#configCell(#token("k","#CellName"),#cellPropertyListTerminator(.KList),#SemanticCastToKItem(#token("$PGM","KConfigVar")),#token("k","#CellName")),#configCell(#token("state","#CellName"),#cellPropertyListTerminator(.KList),`.Map`(.KList),#token("state","#CellName"))),#token("true","Bool"),[org.kframework.attributes.Location(Location(11,19,12,40)), org.kframework.attributes.Source(Source(/home/radu/work/test/test.k)), org.kframework.definition.Production(syntax #RuleContent ::= #RuleBody [klabel(#ruleNoConditions), symbol])])
```

SchmErik replied:
Ah, Thanks! I was kind of looking for something like this. In hindsight, I should have asked you!

Commit:
Use visitor to inspect cell names instead

Commit:
update test results

radumereuta committed:
Merge branch 'master' into check-duplicate-k-cell

radumereuta commented on kernel/src/main/java/org/kframework/kompile/DefinitionParsing.java:
```diff
@@ -339,6 +359,13 @@ private Definition resolveConfigBubbles(Definition def) {
+            KCellDeclarations.stream().forEach(k ->
+                    errors.add(KEMException.compilerError("Multiple K cell declarations detected. Only one declaration is allowed.", k)));
+        }
+
```
Somehow this seems a bit convoluted.
Try this instead:
```java
    private static void checkDuplicateKCells(Definition defWithParsedConfigs) {
        // check for duplicate <k> cell declarations
        List<K> kcells = new ArrayList<>();
        stream(defWithParsedConfigs.mainModule().sentences())
                .filter(s -> s instanceof Configuration)
                .forEach(s -> new VisitK() {
                    @Override
                    public void apply(KApply k) {
                        if (k.klabel().equals(KLabel("#configCell"))) {
                            KToken kt = (KToken) k.klist().items().get(0);
                            assert kt.sort().equals(Sorts.CellName());
                            if (kt.s().equals("k"))
                                kcells.add(k);
                        }
                        super.apply(k);
                    }
                }.apply(((Configuration) s).body()));
        if (kcells.size() > 1)
            throw KEMException.compilerError("Only one <k> cell declaration is allowed. Previously declared at "
                    + kcells.get(0).source().orElse(null) + " and " + kcells.get(0).location().orElse(null), kcells.get(1));
    }
```

radumereuta commented on k-distribution/tests/regression-new/checks/duplicateKCells.k:
```diff
@@ -0,0 +1,12 @@
+module DUPLICATERULES-SYNTAX
```
```suggestion
module DUPLICATEKCELLS-SYNTAX
```

radumereuta commented on k-distribution/tests/regression-new/checks/duplicateKCells.k:
```diff
@@ -0,0 +1,12 @@
+endmodule
+
+module DUPLICATEKCELLS
+    imports DUPLICATERULES-SYNTAX
```
```suggestion
    imports DUPLICATEKCELLS-SYNTAX
```

radumereuta reviewed (changes requested):
I tried to test this but your solution didn't work when the two configurations are in the same module:

```
module TEST
    configuration <env>
                    <k> $PGM:Int </k>
                    <state> .Map </state>
                  </env>
    configuration <k> .K </k>
endmodule
```

or when you have nested <k> cells:
```
    configuration <k>
                    <k> $PGM:Int </k>
                    <state> .Map </state>
                  </k>
```


I added a code snippet that should work.
Please double check.

Commit:
Update k-distribution/tests/regression-new/checks/duplicateKCells.k

Co-authored-by: Radu Mereuta <headness13@gmail.com>

Commit:
Update k-distribution/tests/regression-new/checks/duplicateKCells.k

Co-authored-by: Radu Mereuta <headness13@gmail.com>

Commit:
Applying Radu's feedback

Radu's suggestion is much more concise. The solution to emit an error on
all K cell declarations has been retained because it allows the user to
see all locations of all k cell declarations.

Commit:
Add additional case to test and update results

radumereuta commented on kore/src/main/scala/org/kframework/definition/outer.scala:
```diff
@@ -115,6 +119,8 @@ case class Module(val name: String, val imports: Set[Import], localSentences: Se
   lazy val productions: Set[Production] = sentences collect { case p: Production => p }
 
+  lazy val localConfigurations: Set[Configuration] = localSentences collect { case c: Configuration => c }
+
```
Changes to this file are no longer needed.

radumereuta reviewed (changes requested):
In my version of the code, I throw an exception because of this example:
```
    configuration <k>
                    <k> $PGM:Int </k>
                    <state> .Map </state>
                  </k>
```
The compilation continues for a bit, then fails badly with a cryptic error.

Commit:
Remove unused code

This code was used in a previous solution and no longer needed. Remove
it.

SchmErik commented:
> In my version of the code, I throw an exception because of this example:
> 
> ```
>     configuration <k>
>                     <k> $PGM:Int </k>
>                     <state> .Map </state>
>                   </k>
> ```
> 
> The compilation continues for a bit, then fails badly with a cryptic error.

I tried to kompile this code with the current head and it's also resulting in an exception. We should file a separate issue for this...

radumereuta commented:
> > In my version of the code, I throw an exception because of this example:
> > ```
> >     configuration <k>
> >                     <k> $PGM:Int </k>
> >                     <state> .Map </state>
> >                   </k>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The compilation continues for a bit, then fails badly with a cryptic error.
> 
> I tried to kompile this code with the current head and it's also resulting in an exception. We should file a separate issue for this...

What I was trying to say is that in my code I have
```java
if (kcells.size() > 1)
            throw KEMException.compilerError("Only one <k> cell declaration is allowed. Previously declared at "
                    + kcells.get(0).source().orElse(null) + " and " + kcells.get(0).location().orElse(null), kcells.get(1));
```
This means kompilation is stopped in this spot.
If you do it your way, kompile will run too far and get stuck in some asserts.
We have the opportunity to completely bypass that error with this message.

SchmErik commented:
ah, I see what you mean. I'm fine with taking either solution but the advantage to recording all declarations is that the user can understand where all of the declarations are located and gives them the ability to choose which ones to keep. I wonder if there's a way to do both: throw an exception that contains all of the exceptions...

radumereuta commented:
You could call `private void throwExceptionIfThereAreErrors()`. See usages.

Commit:
Throw exception if multiple K cells are found

There are assertions that cause a crash in later stages of kompile so
abort compilation when multiple K cells are found.

Commit:
update test results

rv-jenkins committed:
Merge branch 'master' into check-duplicate-k-cell
